spec.bs

```diff
@@ -704,15 +694,25 @@ To <dfn>validate and convert auction ad config</dfn> given an {{AuctionAdConfig}
     a {{TypeError}}.
   1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer priority signals=][|buyer|] to
     |signals|.
-1. If |config|["{{AuctionAdConfig/perBuyerCurrencies}}"] [=map/exists=], [=map/for each=]
-  |key| → |value| of |config|["{{AuctionAdConfig/perBuyerCurrencies}}"]:
-  1. If the result of [=checking whether a string is a valid currency tag=] given |value| is false,
-    [=exception/throw=] a {{TypeError}}.
-  1. If |key| is "*", then set |auctionConfig|'s
-    [=auction config/all buyers currency=] to |value|, and [=iteration/continue=].
-  1. Let |buyer| be the result of [=parsing an origin=] with |key|. If |buyer| is failure,
-    [=exception/throw=] a {{TypeError}}.
-  1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer currencies=][|buyer|] to |value|.
+1. If |config|["{{AuctionAdConfig/perBuyerCurrencies}}"] [=map/exists=]:
+  1. Set |auctionConfig|'s [=auction config/per buyer currencies=] to |config|["{{AuctionAdConfig/perBuyerCurrencies}}"]
+  1. Increment |auctionConfig|'s [=auction config/pending promise count=].
+  1. Let |resolvedAndTypeChecked| be the promise representing performing the following steps
+    [=upon fulfillment=] of |auctionConfig|'s [=auction config/per buyer currencies=] with |result|:
+    1. Set |auctionConfig|'s [=auction config/per buyer currencies=] to a new [=ordered map=] whose [=map/keys=] are [=origins=] and whose [=map/values=] are [=currency tags=].
+    1. [=map/for each=] |key| → |value| of |result|:
+      1. If the result of [=checking whether a string is a valid currency tag=] given |value| is false,
+        [=exception/throw=] a {{TypeError}}.
+      1. If |key| is "*", then set |auctionConfig|'s
+        [=auction config/all buyers currency=] to |value|, and [=iteration/continue=].
+      1. Let |buyer| be the result of [=parsing an origin=] with |key|. If |buyer| is failure,
+        [=exception/throw=] a {{TypeError}}.
+      1. [=map/Set=] |auctionConfig|'s [=auction config/per buyer currencies=][|buyer|] to |value|.
+    1. If the previous steps did not [=exception/throw=] an exception:
+      1. Decrement |auctionConfig|'s [=auction config/pending promise count=].
+  1. [=Upon rejection=] of |resolvedAndTypeChecked|:
+    1. Set |auctionConfig|'s [=auction config/per buyer currencies=] to failure.
+    1. Decrement |auctionConfig|'s [=auction config/pending promise count=].

```

Comment:
We're starting to repeat this chunk in a bunch of places. See if you can abstract it into an algorithm, probably one that takes the middle validation steps as a parameter.

Reply:
Good point, done, though of course the syntax may need tweaking.